### ISSUE TITLE
Improvements to exam grading

### ISSUE THREAD

**Issue by rzach** [Carnap/Carnap#81]
List of things I'm noticing while grading exams:

- [x] The view all/view ungraded toggle gives no feedback on what the current view is. The currently in-force option should be highlighted.
- [x] "View all" is a bit confusing. It doesn't show all answers, only those that were incorrect. In https://github.com/gleachkr/Carnap/issues/80#issuecomment-602964104 you mentioned the possibility of assigning extra credit using this feature. But if I can't see solutions the student got right, I can't assign extra credit for especially good solutions.
- [x] Since the UI lets you enter any point value, it would be good to know how many points a problem is worth.
- [x] Credit field should be pre-filled with currently assigned value in the database. (Two reasons: a) let's you just press "update" on a 0 field if you don't want to give partial credit, without an error. b) Changing the value without pressing update is persistent through changes in view and even reloads. This is bad since the value shown is different from what the student is assigned and there's no way to tell)
- [x] Give visual feedback if a credit score has been changed from what's in the database, e.g., by graying out the "update" button if the score in the input box and the database agree. (Should be active for "ungraded" problems.)
- [x] In "view all" there should be some visual indication if a problem has already been assigned points by hand (ie if it is "ungraded" or not)
- [x] Proofs would be easier to grade if we got to see the rendered version instead of he raw input and/or the feedback per line (which inferences are incorrect, etc.)
- [x] Individual students are listed in the drop-down by email, but while this is unique it makes it hard to find a student by name.
~For translation exercises, it's possible to edit the solution provided.  It's unclear if this does anything to the database (eg, can the student see this?). If it serves no purpose to edit the solution it should just be displayed, maybe even in the official parsed form (so the symbols are easier to compare with the correct solution)~
- [x] Sort students alphabetically
- [x] Warn in case of unsaved changes

**Comment by gleachkr** [comment]
About the last one: there might be a purpose to (generally) letting instructors edit submitted answers. Basically, sometimes late at night, it can be nice to get confirmation that you're right when you suspect that a student has made only one small and easily fixable error. An easy way to check is to try to fix the mistake and see if carnap finds anything else.

What do you think? Either way, this should be better documented and more consistently implemented.

**Comment by rzach** [comment]
Oh I see. Yeah, that's actually pretty neat. I didn't think to hit enter. (Maybe we should for uniformity put a "check" button on all types of exercise, then it would be clear here as well that the check for correctness function is available). I could add that above and/or to issue #19.)

**Comment by gleachkr** [comment]
Maybe move to #19?

**Comment by dsanson** [comment]
Another thing I noticed, while grading qualitative short answer problems: there was no good way to provide students with feedback. Ideally, there would be a text box allowing feedback, which would be displayed to the student on the page where they go to check their results. For the moment, I found myself providing feedback by email. For that, it would have been nice to have a  mailto link next to each problem (though that doesn't seem like the best solution long term).

**Comment by gleachkr** [comment]
I wonder if I could get a little bit more detail on 4 and 7? 

For 4, I think it should be doing this already :crossed_fingers:, but is the issue that it should prefill as zero when there's no assigned partial credit?

For 7, is the idea to turn on the `render` option by default when grading?

**Comment by rzach** [comment]
For 4, a 0 if no partial credit is in the database would fix a). For b) I think there is a bug (?) that prepopulates the partial credit field with whatever the browser has in that field, rather than what's in the database (so if you change the partial credit without hitting update and then reloading the page, the changed value is still there.)

For 7, yes, and/or `init=now`.  You can now get the feedback in the proof box but only after you click inside and hit a key. I guess what I wanted was a better display of what's wrong in the proof, or of the proof itself. Proofs are hard to parse esp if there's a lot of subproofs.

**Comment by gleachkr** [comment]
Thanks! I might try a custom initialization procedure to get things to initialize as they scroll into view, in order to avoid locking up the browser with a large number of simultaneous proof checks.

**Comment by gleachkr** [comment]
For 4a, I've changed the prefilling, and added something so that you can see the value in the DB after editing what's in the input field but before submitting. For 4b, I'm not able to reproduce the issue. Is this still affecting you? If so, what browser?

**Comment by gleachkr** [comment]
I've now added that "initialize on scroll into view" behavior to handle 7.

**Comment by rzach** [comment]
Re 4b, on Firefox if I go to grade an assignment, change the point value on a problem, then hit reload, the point value is the one I've put in but haven't saved yet.

I can now tell that it's different than what's in the database (thanks!) but I think reload should reset the page, shouldn't it?

**Comment by gleachkr** [comment]
Ah! OK, reproducible on Firefox. I think it's this:

https://stackoverflow.com/questions/7377301/firefox-keeps-form-data-on-reload

Firefox tries to autocomplete form data by filling in known values. This can be disabled by turing off "autocomplete" on the input. So, did that, and it seems to be fixed. Fix is deploying.

**Comment by rzach** [comment]
It is fixed. Now I'm wondering if there shouldn't be a warning if you've changed a value that you haven't saved.

**Comment by rzach** [comment]
Can we sort the student list in the dropdown alphabetically?

I'm going to take the last point back. I think it's good to be able to edit the solution, since you might want to see if you can fix it by just adding a parenthesis or something.

For item 2 maybe "All/Unreviewed"?

**Comment by gleachkr** [comment]
Yes, that makes sense for 2. I was initially thinking three categories, with "all/not-auto-graded/unreviewed", but that's a bit more complication than might be warranted. I'm not sure I can think of a standard use-case for "not-auto-reviewed" (or a good name for it, evidently).

**Comment by rzach** [comment]
I think it would actually be nice to have a way to see a students answers to questions even if they were right. Maybe better to see them on the student page though rather than here. (Here the only use-case would be to assign extra credit to especially elegant proofs or something, or extra credit for small counterexamples?)

**Comment by gleachkr** [comment]
Agreed. I was thinking that "all" would include both auto-graded and non-auto-graded exercises. But yeah, it would also work to access correct answers through student pages. I'm not sure about use cases. I guess I can imagine it might be pedagogicall helpful to notice that a group of students are turning in word-for-word the same answers, or everybody is solving a proof in a similarly correct but roundabout way.

**Comment by gleachkr** [comment]
I think that's everything! OK to close?

**Comment by rzach** [comment]
Yeah thanks! (Autograded answers other than proofs are not saved, so can't be shown, right?)

**Comment by gleachkr** [comment]
That's right - I think Sequent calculus and Tree ND results are saved too, but not the rest. Maybe it would make sense to save correct translations and countermodels as well, though?

**Comment by rzach** [comment]
Not a high priority (for me) but if you want to be able to give extra credit it would make sense ... also would be nice to see what solutions students come up with. Separate issue?

**Comment by gleachkr** [comment]
Perfect, that'd be great.

**Comment by gleachkr** [comment]
Or sorry, don't mean to presume on your time - I can open it if you're pressed today.

**Comment by rzach** [comment]
LOL presume on *my* time? Thanks for adding that partial credit functionality within hours of me asking for it.
